File: code-ninja-course/JAVA/src/main/java/com/codeninja/solidsolution/students/StudentService.java
package com.codeninja.solidsolution.students;

import com.codeninja.solidsolution.*;
import com.codeninja.solidsolution.exceptions.*;
import com.codeninja.solidsolution.universities.*;

public class StudentService {

	public void addIfValid(	
			String emailAddress, 
			int universityId, 
			StudentRepository studentRepository, 
			UniversityRepository universityRepository ) {
		Logger.log("Start add student with email",emailAddress);

		validateEmail(emailAddress);

		if (studentRepository.exists(emailAddress))	{
			return;
		}

		University university = universityRepository.getByID(universityId);
		Student student = studentRepository.createStudent(emailAddress, universityId);

		PackageType type = university.getPackageType();
		
		BookAllowanceSetter allowanceSetter = new BookAllowanceSetter(student);
		allowanceSetter.setAllowance (type);

		studentRepository.add(student);

		Logger.log("End add student with email", emailAddress);

	}
	
	public Student getStudentsByUniversity() {
		throw new NotImplemented();
	}
	
	public Student getStudentsByCurrentlyBorrowedEbooks() {
		throw new NotImplemented();
	}

	private void validateEmail(String emailAddress) {
		if (emailAddress.isEmpty())
			throw new InvalidEmail("Empty");
	}
}

File: code-ninja-course/CSHARP/CleanCode/Code/SOLID/StudentService.cs
using System;

namespace CleanCode
{
    public class StudentService
    {

        public bool addIfValid(string emailAddress, int universityId)
        {
            //SRP
            Console.Write("Log: Start add student with email %s", emailAddress);

            if (String.IsNullOrEmpty(emailAddress))
            {
                return false;
            }

            StudentRepository studentRepository =
                    new StudentRepository();

            if (studentRepository.exists(emailAddress))
            {
                return false;
            }

            UniversityRepository universityRepository =
                    new UniversityRepository();

            University university =
                    universityRepository.getByID(universityId);

            Student student = new Student(emailAddress, universityId);

            //SRP, OCP
            PackageType type =
                      university.GetPackageType();
            if (type == PackageType.Standard)
            {
                student.setMonthlyEbookAllowance(10);
            }
            else if (type == PackageType.Premium)
            {
                student.setMonthlyEbookAllowance(10 * 2);
            }

            studentRepository.add(student);

            //SRP
            Console.Write("Log: End add student with email %s", emailAddress);

            return true;//CQS

        }

        public Student getStudentsByUniversity()
        {
            throw new NotImplementedException();
        }

        public Student getStudentsByCurrentlyBorrowedEbooks()
        {
            throw new NotImplementedException();
        }
    }

}
